File: Editorial/123_Best_Time_to_Buy_and_Sell_Stock_III/CSharp/2_One_pass_Simulation.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        int t1Cost = Int32.MaxValue, t2Cost = Int32.MaxValue;
        int t1Profit = 0, t2Profit = 0;
        foreach (int price in prices) {
            // the maximum profit if only one transaction is allowed
            t1Cost = Math.Min(t1Cost, price);
            t1Profit = Math.Max(t1Profit, price - t1Cost);
            // reinvest the gained profit in the second transaction
            t2Cost = Math.Min(t2Cost, price - t1Profit);
            t2Profit = Math.Max(t2Profit, price - t2Cost);
        }

        return t2Profit;
    }
}

File: Editorial/123_Best_Time_to_Buy_and_Sell_Stock_III/Go/2_One_pass_Simulation.go
func maxProfit(prices []int) int {
    t1Cost := math.MaxInt32
    t2Cost := math.MaxInt32
    t1Profit := 0
    t2Profit := 0
    for _, price := range prices {
        // the maximum profit if only one transaction is allowed
        t1Cost = int(math.Min(float64(t1Cost), float64(price)))
        t1Profit = int(math.Max(float64(t1Profit), float64(price-t1Cost)))
        // reinvest the gained profit in the second transaction
        t2Cost = int(math.Min(float64(t2Cost), float64(price-t1Profit)))
        t2Profit = int(math.Max(float64(t2Profit), float64(price-t2Cost)))
    }
    return t2Profit
}
